core/src/main/java/org/wso2/carbon/testgrid/core/TestGridMgtServiceImpl.java

```diff
@@ -83,27 +82,29 @@
         return infras;
     }
 
-    private List<TestPlan> generateTestPlan(String repoDir, String homeDir) throws TestGridException {
-        String productTestPlanDir = repoDir + File.separator + PRODUCT_TEST_DIR;
+    private CopyOnWriteArrayList<TestPlan> generateTestPlan(String testRepoDir, String infraRepoDir, String homeDir) throws
+            TestGridException {
+        String productTestPlanDir = testRepoDir + File.separator + PRODUCT_TEST_DIR;
         File dir = new File(productTestPlanDir);
         File[] directoryListing = dir.listFiles();
-        List<TestPlan> testPlanList = new ArrayList<>();
+        CopyOnWriteArrayList<TestPlan> testPlanList = new CopyOnWriteArrayList<>();
```

Comment:
Should use `List<TestPlan> testPlanList = new CopyOnWriteArrayList<>();` since we need to programme to the interface [1]

[1] https://dzone.com/articles/programming-to-an-interface

Reply:
Fixed with 14598a57750f800c12e6cb5a3828474056e7535c.